### ISSUE TITLE
`GenericArray` should be `#[repr(transparent)]`

### ISSUE THREAD

**Issue by Ekleog** [fizyk20/generic-array#74]
This would make it possible to have a `mem::uninitialized()` `GenericArray` without insta-UB.

Thank you for your work on generic-array! :)

**Comment by oconnor663** [comment]
There are also unsafe pointer casts to `GenericArray` in the `From` impls, and I think that's relying on unspecified type layout without either `#[repr(transparent)]` or `#[repr(C)]`.

**Comment by novacrazy** [comment]
I've actually tried as hard as I can to get UB on `GenericArray` as it is [using exotic element sizes and packing](https://github.com/fizyk20/generic-array/blob/master/tests/mod.rs#L179), but it never had any issues. Maybe on something other than x86_64 it would.

However, it does make sense for `GenericArray` itself to be `#[repr(transparent)]` as it only has the single value inside it.

Previously that was not even an option, as `#[repr(transparent)]` was only stabilized in 1.28.0, and `generic-array` needed to be compatible for 1.21.0... However, with `generic-array` 0.13.0, we've bumped the minimum required Rust version to 1.31.0, so this is totally doable immediately.
